docs/arch/adr-001-send-unpublish-emails-for-single-pages.md

```diff
@@ -0,0 +1,61 @@
+# Decision record: reinstate unpublish emails for single page subscriptions
+
+Date: 2021-12-03
+
+## Context
+
+We have just launched a new email notifications feature that allows users to subscribe to updates on a single page.
+Currently this feature is only enabled on a few selected pages (eg. [Open standards for government][open-standards-for-government]),
+but it is expected to roll out to the majority of pages and content types by early 2022.
+There is a user need to let subscribers know when a page they are subscribed to has been withdrawn or redirected.
+
+The email alert API used to send emails when a page in a taxon subscription was unpublished.
+This feature was [removed in Jan 2021][remove-previous-unpublish-emails] for good reasons.
+
+[open-standards-for-government]: https://www.gov.uk/government/publications/open-standards-for-government
+[remove-previous-unpublish-emails]: https://github.com/alphagov/email-alert-api/issues/1572
+
+## Decision
+
+We will reinstate emails when a page is unpublished, but only to users with a single page subscription to that page.
+We will also remove the subscriber list for a single page when it is unpublished to reduce the build up of empty lists in Email Alert API.
```

Comment:
It'd be good to explain a little about what type of endpoint Email Alert API will have for this, it seems a tricky decision modelling/terminology wise and one that'd be good to have thought put into. It'll be quite a strange thing to add in nicely because Email Alert API doesn't have an understanding of a publishing workflow. 

Some things, with various pros and cons, that come to mind are:

- a destroy endpoint for subscriber lists - a rather restful approach that adds a generic endpoint to Email Alert API to end a subscriber list and it can optionally send an email. Has pros of not adding an extra domain into Email Alert API; has downsides that you'd have to work out how to get the right subscriber list from Email Alert Service and compose the copy there.
- a single page unpublished endpoint - an endpoint that has the sole purpose of notifying users for this feature and would be called each time any page is unpublished. It'd have the responsibility of mapping publishing workflow to lists. If any extra unpublishing behaviour was added to Email Alert API another endpoint would be needed. This would be very clear in purpose but feel rather specific
- a generic unpublished endpoint that actually only does one thing, single pages - this would potentially be the most confusing of scenarios where we have an endpoint that looks like it might do a lot but it actually only does a small bit these single pages and may linger for a long time as something that someone one day may iterate to do more but never happens.

For context, these are probably ordered in my preference order from cleanest to messiest integration

Reply:
I like (1).

Unfortunately, it looks like request bodies for DELETE requests [are a bit questionable](https://stackoverflow.com/questions/299628/is-an-entity-body-allowed-for-an-http-delete-request) - technically permitted but might not work in all cases.  So probably the best way to do it would be to have a route:

```ruby
post "/subscriber-lists/:slug/delete", to "subscriber_lists#destroy"
```

Which by default unsubscribes everyone from the list and destroys it *without* sending an email.

To send an email, it would have an optional parameter `body_markdown`.  @alex9smith I see in the email-alert-api branch you also want `url` and `title` parameters: can't you just take those from the `SubscriberList`?

I think this is nicer than the current `create_unpublishing_notification` endpoint in the draft email-alert-api PR, as that branch is (1) very specific to single-page notifications

Reply:
> We will also remove the subscriber list for a single page when it is unpublished to reduce the build up of empty lists in Email Alert API.

If we end the subscriptions we don't actually need to do this, because of the `HistoricalDataDeletionWorker`.

Reply:
Perhaps this change would work for email-alert-api:

```diff
diff --git a/app/controllers/subscriber_lists_controller.rb b/app/controllers/subscriber_lists_controller.rb
index a3f79edb..5ce08fe5 100644
--- a/app/controllers/subscriber_lists_controller.rb
+++ b/app/controllers/subscriber_lists_controller.rb
@@ -31,6 +31,22 @@ class SubscriberListsController < ApplicationController
     render json: subscriber_list.to_json
   end
 
+  def destroy
+    subscriber_list = SubscriberList.find_by(slug: params[:slug])
+    if subscriber_list
+      DestroySubscriberListService.call(
+        subscriber_list: subscriber_list,
+        params: params.permit(:title, :body).to_h,
+        user: current_user,
+        govuk_request_id: GdsApi::GovukHeaders.headers[:govuk_request_id],
+      )
+
+      head :no_content
+    else
+      render json: { error: "Could not find the subscriber list" }, status: :not_found
+    end
+  end
+
 private
 
   def convert_legacy_params(link_or_tags)
diff --git a/app/models/subscription.rb b/app/models/subscription.rb
index b9eb14f8..50506605 100644
--- a/app/models/subscription.rb
+++ b/app/models/subscription.rb
@@ -21,7 +21,7 @@ class Subscription < ApplicationRecord
     frequency_changed: 2,
     subscriber_list_changed: 3,
     marked_as_spam: 4,
-    unpublished: 5, # Unused since 5eeda132 (can be removed after a year)
+    unpublished: 5,
     bulk_immediate_to_digest: 6, # Potentially unused (for a one-off migration)
     subscriber_merged: 7,
   }, _prefix: :ended
diff --git a/app/queries/subscriber_lists_by_criteria_query.rb b/app/queries/subscriber_lists_by_criteria_query.rb
index 54c6d2b3..cacab663 100644
--- a/app/queries/subscriber_lists_by_criteria_query.rb
+++ b/app/queries/subscriber_lists_by_criteria_query.rb
@@ -19,6 +19,7 @@ private
   attr_reader :initial_scope, :criteria_rules
 
   def rule_condition(scope, rule)
+    return scope.where(id: rule[:id]) if rule[:id]
     return type_rule(scope, **rule) if rule[:type]
     return any_of_rule(scope, rule[:any_of]) if rule[:any_of]
     return all_of_rule(scope, rule[:all_of]) if rule[:all_of]
diff --git a/app/services/destroy_subscriber_list_service.rb b/app/services/destroy_subscriber_list_service.rb
new file mode 100644
index 00000000..53c4472c
--- /dev/null
+++ b/app/services/destroy_subscriber_list_service.rb
@@ -0,0 +1,33 @@
+class DestroySubscriberListService
+  include Callable
+
+  attr_reader :subscriber_list, :params, :govuk_request_id, :user
+
+  def initialize(subscriber_list:, params:, govuk_request_id:, user:)
+    @subscriber_list = subscriber_list
+    @params = params
+    @govuk_request_id = govuk_request_id
+    @user = user
+  end
+
+  def call
+    message = Message.create!(message_params) if message_params
+    Metrics.message_created if message
+    ProcessDestroySubscriberListWorker.perform_async(
+      subscriber_list.id,
+      message&.id,
+    )
+  end
+
+  def message_params
+    return unless params[:title] && params[:body]
+
+    params
+      .slice(:title, :body)
+      .merge(
+        criteria_rules: { id: subscriber_list.id },
+        govuk_request_id: govuk_request_id,
+        signon_user_id: signon_user_id,
+      )
+  end
+end
diff --git a/app/workers/process_destroy_subscriber_list_worker.rb b/app/workers/process_destroy_subscriber_list_worker.rb
new file mode 100644
index 00000000..60cf1f80
--- /dev/null
+++ b/app/workers/process_destroy_subscriber_list_worker.rb
@@ -0,0 +1,14 @@
+class ProcessDestroySubscriberListWorker < ApplicationWorker
+  sidekiq_options queue: :process_and_generate_emails
+
+  def perform(subscriber_list_id, message_id)
+    run_with_advisory_lock(SubscriberList, subscriber_list_id) do
+      ProcessMessageWorker.new.perform(message_id) if message_id
+
+      Subscription.active.where(subscriber_list_id: subscriber_list_id).update_all!(
+        ended_reason: :unpublished,
+        ended_at: Time.zone.now,
+      )
+    end
+  end
+end
diff --git a/config/routes.rb b/config/routes.rb
index 92955b81..7af1d50f 100644
--- a/config/routes.rb
+++ b/config/routes.rb
@@ -4,6 +4,7 @@ Rails.application.routes.draw do
     resources :subscriber_lists, path: "subscriber-lists", only: %i[create]
     get "/subscriber-lists", to: "subscriber_lists#index"
     get "/subscriber-lists/:slug", to: "subscriber_lists#show"
+    post "/subscriber-lists/:slug/delete", to: "subscriber_lists#destroy"
 
     resources :content_changes, only: %i[create], path: "content-changes"
     resources :messages, only: %i[create]
```

Adjusting the `Message` machinery to support matching a specific `SubscriberList` is a tad clunky, but feels better than implementing essentially the same feature a second time.

Reply:
I've opened a draft PR: https://github.com/alphagov/email-alert-api/pull/1694

Reply:
Will resolve this here and move any further chat over to the PR